### PR TITLE
[2/n] Add sparse softmax to the Triton flash attention kernel

### PR DESCRIPTION
**Pull request by kaix-nv**
### What does this PR do?

Type of change: ? 

Type of change: New feature

Add N:M structured sparsity support to the Triton flash attention kernel (`modelopt/torch/kernels/triton_fa.py`). For every M consecutive key positions in the attention score tile, keeps the top-N values and sets the rest to -inf before softmax. This is applied during prefill only.

**Supported patterns:** Any N:M where M=4 (N=1,2,3) or M=8 (N=1..4).

- Sink tokens and dense window blocks for preserving local attention and attention sinks

 **Performance (TFLOPS at seq_len=16384, RTX 6000):**
  | Pattern | TFLOPS | % of Dense |
  |---------|--------|------------|
  | Dense | 89.3 | 100% |
  | 2:4 (M=4) | 69.5 | 78% |
  | 4:8 (M=8) | 57.3 | 64% |

### Usage

```python
# Add a code snippet demonstrating how to use this
from modelopt.torch.kernels import attention

# 2:4 sparsity (keep top 2 of every 4 K positions)
out = attention(q, k, v, b_start_loc, b_seq_len, max_len,
                sparsity_n=2, sparsity_m=4)

# 4:8 sparsity with sink tokens and dense window
out = attention(q, k, v, b_start_loc, b_seq_len, max_len,
                sparsity_n=4, sparsity_m=8,
                num_sink_tokens=4, dense_window_blocks=2)

# Dense (default, zero overhead)
out = attention(q, k, v, b_start_loc, b_seq_len, max_len)

# Via mtsa.sparsify() on HuggingFace models
import modelopt.torch.sparsity.attention_sparsity as mtsa
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.1-8B",
                                              torch_dtype=torch.bfloat16,
                                              device_map="cuda")

# Default config
mtsa.sparsify(model, mtsa.SPARSE_SOFTMAX_DEFAULT)
```

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / ❌ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * N:M structured sparse softmax for Triton flash-attention prefill with configurable dense-window and sink-token handling.

* **API**
  * attention(...) accepts sparsity_n, sparsity_m, num_sink_tokens, dense_window_size; HF/Triton prefill path propagates them.

* **Configuration**
  * New config fields and exported preset to enable/configure Triton N:M sparse softmax with validation.

* **Tests**
  * Added GPU tests covering N:M behavior, tile structure, forward/backward correctness.

* **Documentation**
  * CHANGELOG and example docs updated with usage and CLI options.

### CODE REVIEW

File: modelopt/torch/sparsity/attention_sparsity/methods/triton_sparse_softmax.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    def name(self) -> str:
+        """Method name identifier."""
+        return "triton_sparse_softmax"
+
+    def calculate_sparsity(self, attention_scores):
+        """Return a no-op mask (sparsity is applied inside the Triton kernel)."""
+        mask = torch.ones_like(attention_scores, dtype=torch.bool)
+        return mask, {}
+
+    def apply_sparsity(self, attention_scores, sparse_mask=None):
```
Comment (rohansjoshi) [inline modelopt/torch/sparsity/attention_sparsity/methods/triton_sparse_softmax.py]: It would be better if, for Triton-backend attention sparsity methods, we don't have to add the boilerplate in this file (i.e. not have to add a new file to methods/ altogether)

**Comment by kaix-nv** [comment]
@coderabbitai full review

File: modelopt/torch/sparsity/attention_sparsity/methods/triton_sparse_softmax.py (hunk 1 above)
Comment (kaix-nv) [reply modelopt/torch/sparsity/attention_sparsity/methods/triton_sparse_softmax.py]: Good catch. I've made `calculate_sparsity` and `apply_sparsity` optional in base class.

File: modelopt/torch/kernels/hf_triton_attention.py [hunk 2]
```diff
@@ -105,6 +105,15 @@ def triton_attention_forward(
         kw["b_seq_len_k"] = torch.full((batch,), seq_k, device=device, dtype=torch.int32)
         kw["max_input_len_k"] = seq_k
 
+    # N:M sparse softmax — prefill only (decode should not sparsify KV)
+    if not is_decode and getattr(module, "_apply_sparse_nm", False):
+        method = getattr(module, "_sparse_method_instance", None)
```
Comment (Edwardf0t1) [inline modelopt/torch/kernels/hf_triton_attention.py:112]: Where `_sparse_method_instance` gets set, if it's outside this PR, please add a comment.

File: modelopt/torch/kernels/triton_fa.py [hunk 3]
```diff
@@ -162,6 +276,21 @@ def _attn_fwd(
...
+            is_sink = kv_start < NUM_SINK_TOKENS
+            # causal_offset handles chunked prefill: q starts at (seq_len_kv - seq_len_q)
+            causal_offset = seq_len_kv - seq_len_q
+            q_abs_pos = tile_q * BLOCK_M + causal_offset
+            token_distance = q_abs_pos - kv_start
+            is_local = (token_distance >= 0) and (token_distance < DENSE_WINDOW_SIZE)
+            if not is_sink and not is_local:
+                scores = _apply_sparse_nm_to_qk_tile(
+                    scores, BLOCK_M, BLOCK_N, SPARSITY_N, SPARSITY_M
+                )
```
Comment (Edwardf0t1) [inline modelopt/torch/kernels/triton_fa.py]: Consider extracting the duplicated sink/window check into a shared helper function.

File: modelopt/torch/kernels/hf_triton_attention.py (hunk 2 above)
Comment (kaix-nv) [reply modelopt/torch/kernels/hf_triton_attention.py:112]: Added

File: modelopt/torch/kernels/triton_fa.py (hunk 3 above)
Comment (kaix-nv) [reply modelopt/torch/kernels/triton_fa.py]: Good catch, fixed.

**Comment by kaix-nv** [comment]
@coderabbitai full review
